File: week-03_05/tp/tp/tp4/solution/transpose_resized.f90
PROGRAM transpose
  USE MPI_F08
  IMPLICIT NONE
  INTEGER, PARAMETER                       :: nb_lines=5,nb_columns=4,&
                                              nb_lines_t=4,nb_columns_t=5,&
                                              tag=1000
  INTEGER                                  :: rank,size_real,i
  TYPE(MPI_Datatype)                       :: type_line,type_transpose
  REAL, DIMENSION(nb_lines,nb_columns)     :: A
  REAL, DIMENSION(nb_lines_t,nb_columns_t) :: AT
  INTEGER(kind=MPI_ADDRESS_KIND)           :: new_size, new_lb=0
  TYPE(MPI_Status)                         :: status

  !-- MPI init 
  CALL MPI_INIT()

  !-- Who I am
  CALL MPI_COMM_RANK(MPI_COMM_WORLD,rank)

  !-- AT matrix  initialisation 
  AT(:,:) = 0.

  !Size of the generic datatype MPI_REAL
  CALL MPI_TYPE_SIZE(MPI_REAL,size_real)

  !Creation of the derived datatype type_line 
  !for one line of the A matrix
  CALL MPI_TYPE_VECTOR(nb_columns,1,nb_lines,MPI_REAL,type_line)

  !Creation of the derived datatype type_transpose to transpose the
  !A matrix nb_lines lines and nb_columns columns
  new_size=size_real
  CALL MPI_TYPE_CREATE_RESIZED(type_line,new_lb,new_size,&
                               type_transpose)

  !Validation of the derived datatype type_transpose
  CALL MPI_TYPE_COMMIT(type_transpose)

  IF (rank == 0) THEN
    !Initialisation of the A matrix on the 0 process
    A(:,:) = RESHAPE( (/ (i,i=1,nb_lines*nb_columns) /), &
                      (/ nb_lines,nb_columns /) )
    PRINT *,'Matrice A'
    DO i=1,nb_lines
      PRINT *,A(i,:)
    END DO
    !Send A matrix to process 1 using the derived datatype type_transpose
    CALL MPI_SEND(A,nb_lines,type_transpose,1,tag,MPI_COMM_WORLD)
  ELSE
    !Reception for the process 1 in AT matrix
    CALL MPI_RECV(AT,nb_lines_t*nb_columns_t,MPI_REAL,0,tag,&
                  MPI_COMM_WORLD,status)
    PRINT *,'Matrix transpose AT'
    DO i=1,nb_lines_t
      PRINT *,AT(i,:)
    END DO
  END IF

  ! Clean MPI types
  CALL MPI_TYPE_FREE(type_transpose)
  CALL MPI_TYPE_FREE(type_line)
  !MPI finalize
  CALL MPI_FINALIZE()
END PROGRAM transpose

File: week-03_05/tp/tp_C/tp4/solution/transpose_resized.c
#include "mpi.h"
#include <stdlib.h>
#include <stdio.h>

int main(int argc, char *argv[]) {
  int nb_lines=4;
  int nb_columns=5;
  int nb_lines_t=5;
  int nb_columns_t=4;
  int tag=1000;
  double a[nb_lines][nb_columns];
  double at[nb_lines_t][nb_columns_t];
  int rank,iterl,iterc,size_real;
  MPI_Datatype type_column, type_transpose;
  MPI_Aint new_extent,new_lbound=0;
  MPI_Status status;

  /* MPI Init */
  MPI_Init( &argc, &argv);

  /* Who I am */
  MPI_Comm_rank( MPI_COMM_WORLD, &rank);

  /* type_column for at */
  MPI_Type_vector(nb_lines_t,1,nb_columns_t,MPI_DOUBLE,&type_column);
  /* Type_transpose for at */
  MPI_Type_size(MPI_DOUBLE,&size_real);
  new_extent = size_real;
  MPI_Type_create_resized(type_column,new_lbound,new_extent,
                          &type_transpose);
  /* Validation of the derived datatypes type_transpose */
  MPI_Type_commit(&type_transpose);

  if (rank == 0) {
    /* Initialisation of A */
    for (iterl=0; iterl<nb_lines; iterl++)
      for (iterc=0; iterc<nb_columns; iterc++)
        a[iterl][iterc] = 1+iterl*nb_columns+iterc;
    printf("Matrix a\n");
    for (iterl=0; iterl<nb_lines;iterl++) {
      for (iterc=0; iterc<nb_columns; iterc++) {
        printf("%4.f ", a[iterl][iterc]);
      }
      printf("\n");
    }

    /* Send matrix A to process 1 */
    MPI_Send(a,nb_columns*nb_lines,MPI_DOUBLE,1,tag,MPI_COMM_WORLD);
  } else {
    /* Receive in matrix AT with type_transpose */
    MPI_Recv(at,nb_columns_t,type_transpose,0,tag,
             MPI_COMM_WORLD,&status);

    printf("Matrice transposee at\n");
    for (iterc=0; iterc<nb_lines_t; iterc++) {
      for (iterl=0; iterl<nb_columns_t;iterl++) {
        printf("%4.f ", at[iterc][iterl]);
      }
      printf("\n");
    }
  }

  /* Clean MPI types */
  MPI_Type_free(&type_column);
  MPI_Type_free(&type_transpose);
  /* Finalize MPI */
  MPI_Finalize();
  return 0;
}
